<!doctype html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">

<!-- Begin Jekyll SEO tag v2.8.0 -->
<title>ssGSEA (v10.1.x) | ssGSEA GenePattern Module Documentation</title>
<meta name="generator" content="Jekyll v3.9.2" />
<meta property="og:title" content="ssGSEA (v10.1.x)" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="A GenePattern module for running the ssGSEA method" />
<meta property="og:description" content="A GenePattern module for running the ssGSEA method" />
<link rel="canonical" href="https://gsea-msigdb.github.io/ssGSEA-gpmodule/v10/" />
<meta property="og:url" content="https://gsea-msigdb.github.io/ssGSEA-gpmodule/v10/" />
<meta property="og:site_name" content="ssGSEA GenePattern Module Documentation" />
<meta property="og:type" content="website" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="ssGSEA (v10.1.x)" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"WebPage","description":"A GenePattern module for running the ssGSEA method","headline":"ssGSEA (v10.1.x)","url":"https://gsea-msigdb.github.io/ssGSEA-gpmodule/v10/"}</script>
<!-- End Jekyll SEO tag -->

    <link rel="stylesheet" href="/ssGSEA-gpmodule/assets/css/style.css?v=a882e2a5706f9c87c0e637123255f61cb5510e46">
    <script src="https://code.jquery.com/jquery-3.3.0.min.js" integrity="sha256-RTQy8VOmNlT6b2PIRur37p6JEBZUE7o8wPgMvu18MC4=" crossorigin="anonymous"></script>
    <script src="/ssGSEA-gpmodule/assets/js/main.js"></script>
    <!--[if lt IE 9]>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv.min.js" integrity="sha256-3Jy/GbSLrg0o9y5Z5n1uw0qxZECH7C6OQpVBgNFYa0g=" crossorigin="anonymous"></script>
    <![endif]-->
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

  </head>
  <body>

      <header>
        <h1>ssGSEA GenePattern Module Documentation</h1>
        <p> A GenePattern module for running the ssGSEA method</p>
      </header>

    <div class="wrapper">
      <nav>
        <ul><b>TABLE OF CONTENTS</b></br></br></ul>
      </nav>
      <section>
        <h1 id="ssgsea-v101x">ssGSEA (v10.1.x)</h1>

<p>Performs single sample GSEA</p>

<p><strong>Author:</strong> GenePattern</p>

<p><strong>Contact:</strong></p>

<p><a href="http://software.broadinstitute.org/cancer/software/genepattern/contact">Contact the GenePattern
team</a>
for GenePattern issues.</p>

<p><a href="https://groups.google.com/forum/#!forum/gsea-help">See the GSEA forum</a>
for GSEA questions.</p>

<p><strong>Algorithm Version:</strong></p>

<h2 id="description">Description</h2>

<p>Project each sample within a data set onto a space of gene set
enrichment scores using the ssGSEA projection methodology described in
Barbie et al., 2009.</p>

<h2 id="summary">Summary</h2>

<p>Single-sample GSEA (ssGSEA), an extension of Gene Set Enrichment
Analysis (GSEA), calculates separate enrichment scores for each pairing
of a sample and gene set. Each ssGSEA enrichment score represents the
degree to which the genes in a particular gene set are coordinately up-
or down-regulated within a sample.</p>

<p><strong>NOTE:</strong> with the release of v10.0.1, this module was renamed from
“ssGSEAProjection” to just “ssGSEA”.</p>

<h2 id="discussion">Discussion</h2>

<p>When analyzing genome-wide transcription profiles from microarray data,
a typical goal is to find genes significantly differentially correlated
with distinct sample classes defined by a particular phenotype (e.g.,
tumor vs. normal). These findings can be used to provide insights into
the underlying biological mechanisms or to classify (predict the
phenotype of) a new sample. Gene Set Enrichment Analysis (GSEA)
addressed this problem by evaluating whether a priori defined sets of
genes, associated with particular biological processes (such as
pathways), chromosomal locations, or experimental results are enriched
at either the top or bottom of a list of differentially expressed genes
ranked by some measure of differences in a gene’s expression across
sample classes. Examples of ranking metrics are fold change for
categorical phenotypes (e.g., tumor vs. normal) and Pearson correlation
for continuous phenotypes (e.g., age). Enrichment provides evidence for
the coordinate up- or down-regulation of a gene set’s members and the
activation or repression of some corresponding biological process.</p>

<p>Where GSEA generates a gene set’s enrichment score with respect to
phenotypic differences across a collection of samples within a dataset,
ssGSEA calculates a separate enrichment score for each pairing of sample
and gene set, independent of phenotype labeling. In this manner, ssGSEA
transforms a single sample’s gene expression profile to a <em>gene
set</em> enrichment profile. A gene set’s enrichment score represents the
activity level of the biological process in which the gene set’s members
are coordinately up- or down-regulated. This transformation allows
researchers to characterize cell state in terms of the activity levels
of biological processes and pathways rather than through the expression
levels of individual genes.</p>

<p>In working with the transformed data, the goal is to find biological
processes that are differentially active across the phenotype of
interest and to use these measures of process activity to characterize
the phenotype. Thus, the benefit here is that the ssGSEA projection
transforms the data to a higher-level (pathways instead of genes) space
representing a more biologically interpretable set of features on which
analytic methods can be applied.</p>

<p>As a practical matter, ssGSEA essentially reduces the dimensionality of
the set. You can look for correlations between the gene set enrichment
scores and the phenotype of interest (e.g., tumor vs. normal) using the
GCT output with a module like ComparativeMarkerSelection. You
could also try clustering the data set; whichever gene sets stand out
as strong predictors of the phenotype of interest, specific clusters can
then be mapped to biochemical pathways, giving you insight into what is
driving the phenotype of interest.</p>

<p>While the GCT can be passed along to any module accepting that format,
it does not make sense to run it through GSEA.</p>

<p>This module implements the single-sample GSEA projection methodology
described in Barbie et al, 2009.</p>

<p><strong>NOTE:</strong> with the release of v10.0.1, this module was renamed from
“ssGSEAProjection” to just “ssGSEA” for clarity and brevity, as it is
commonly referred to by this name. The underlying algorithm and code
remain the same.</p>

<h2 id="references">References</h2>

<ol>
  <li>Subramanian A, Tamayo P, Mootha VK, Mukherjee S, Ebert BL, Gillette
MA, Paulovich A, Pomeroy SL, Golub TR, Lander ES, Mesirov JP. Gene
set enrichment analysis: A knowledge-based approach for interpreting
genome-wide expression profiles. <em>PNAS</em>. 2005;102(43):15545-15550.
<a href="http://www.pnas.org/content/102/43/15545.abstract">http://www.pnas.org/content/102/43/15545.abstract</a></li>
  <li>Barbie DA, Tamayo P, et al. Systematic RNA interference reveals that
oncogenic KRAS-driven cancers require TBK1. <em>Nature</em>.
2009;462:108-112.</li>
</ol>

<h2 id="parameters">Parameters</h2>

<ul>
  <li><strong>Input gct file</strong><span style="color:red;">*</span>
    <ul>
      <li><a href="&quot;http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GCT:_Gene_Cluster_Text_file_format_.28.2A.gct.29&quot;">GCT</a> file containing input dataset’s gene expression data.</li>
    </ul>
  </li>
  <li><strong>Output file prefix</strong>
    <ul>
      <li>The prefix used for the name of the output GCT file. If unspecified, output prefix will be set to &lt;prefix of input GCT file&gt;.PROJ. The output GCT file will contain the projection of input dataset onto a space of gene set enrichments scores.</li>
    </ul>
  </li>
  <li><strong>Gene sets database files</strong><span style="color:red;">*</span>
    <ul>
      <li>This parameter’s drop-down allows you to select gene sets from the <a href="http://www.gsea-msigdb.org/gsea/msigdb/index.jsp">Molecular Signatures Database (MSigDB)</a> on the GSEA website. This drop-down provides access to only the most current version of MSigDB. You can also upload your own gene set file(s) in <a href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29">GMT</a>, <a href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMX:_Gene_MatriX_file_format_.28.2A.gmx.29">GMX</a>, or <a href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GRP:_Gene_set_file_format_.28.2A.grp.29">GRP</a> format.</li>
      <li>If you want to use files from an earlier version of MSigDB you will need to download them from the archived releases on the <a href="http://www.gsea-msigdb.org/gsea/downloads.jsp">website</a>.</li>
    </ul>
  </li>
  <li><strong>Gene symbol column</strong><span style="color:red;">*</span>
    <ul>
      <li>Input GCT file column containing gene symbol names. In most cases, this will be column 1. (default: <em>Column 1</em>)</li>
    </ul>
  </li>
  <li><strong>Gene set selection</strong><span style="color:red;">*</span>
    <ul>
      <li>Comma-separated list of gene set names on which to project the input expression data. Alternatively, this field may be set to <em>ALL</em>, indicating that the input expression dataset is to be projected to all gene sets defined in the specified gene set database(s). (default: <em>ALL</em>)</li>
    </ul>
  </li>
  <li><strong>Sample normalization method</strong><span style="color:red;">*</span>
    <ul>
      <li>Normalization method applied to expression data (if the option <em>“none”</em> is selected, data will not be normalized). Supported methods are <em>none</em>, <em>rank</em>, <em>log.rank</em>, and <em>log</em>. (Default: <em>none</em>)</li>
    </ul>
  </li>
  <li><strong>Weighting exponent</strong><span style="color:red;">*</span>
    <ul>
      <li>Exponential weight employed in calculation of enrichment scores. The default value of 0.75 was selected after extensive testing. The module authors strongly recommend against changing from default. (Default: <em>0.75</em>)</li>
    </ul>
  </li>
  <li><strong>Min gene set size</strong><span style="color:red;">*</span>
    <ul>
      <li>Exclude from the projection gene sets whose overlap with the genes listed in the input GCT file are less than this value. (Default: 10)</li>
    </ul>
  </li>
  <li><strong>Combine mode</strong><span style="color:red;">*</span>
    <ul>
      <li>Options for combining enrichment scores for paired _UP and _DN gene sets. (Default: <em>combine.add</em>)</li>
      <li>For gene set collections that do not utilize _UP and _DN suffixes at the ends of set names, the <em>combine mode</em> parameter option is irrelevant as all the modes give the same output.</li>
      <li>For Gene set collections that utilize _UP and _DN suffixes, which include MSigDB v5’s <em>C2.all</em>, <em>C2.CGP</em>, <em>C6.all</em>, and <em>C7.all</em>, recombine sets in two different ways:
        <ul>
          <li><em>Combine.add</em> mode [default] keeps the original paired sets with independent scores and adds the new combined set with combined score to the output file, increasing the number of gene sets by the number of paired sets.</li>
          <li><em>Combine.replace</em> replaces the original paired gene sets with a new combined gene set and thus reduces the number of gene sets in the output by half the paired sets. Combined sets report combined enrichment scores.</li>
          <li><em>Combine.off</em> keeps _UP and _DN sets separate.</li>
        </ul>
      </li>
    </ul>
  </li>
</ul>

<p><span style="color:red;">*</span> = required</p>

<h2 id="input-files">Input Files</h2>

<ol>
  <li>Input expression dataset</li>
</ol>

<p>The GCT file containing the input dataset’s gene expression data
(see <a href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GCT:_Gene_Cluster_Text_file_format_.28.2A.gct.29">the GCT file format information</a>).
Gene symbols are typically listed in the column with header <em>Name</em>;
however, GCT files containing RNAi data may list the gene symbol
name in alternative columns. The “gene symbol column name”
parameter specifies which of the input GCT file’s columns contains
the gene symbols.</p>

<p>The input GCT file’s row identifiers must draw from the same family
of gene identifiers (the same ontology or name space, such as HUGO
Gene Nomenclature) as those used to identify genes in the gene sets
database file (see next item below). Typically these are human gene
symbols.</p>

<p>If a GCT file’s row identifiers are probe IDs, and gene sets are
defined through a list of human gene symbols, it will be necessary
to collapse all probe set expression values for a given gene into a
single expression value and use a human gene symbol to represent
that gene. The CollapseDataset GenePattern module can make this
transformation.</p>

<ol>
  <li>Gene sets database files</li>
</ol>

<p>One or more optional GMT or GMX file containing a collection of gene
set definitions (see <a href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29">the GMT file
format</a> and 
<a href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMX:_Gene_MatriX_file_format_.28.2A.gmx.29">the GMX file format</a> in the GenePattern file formats documentation).</p>

<h2 id="output-files">Output Files</h2>

<ol>
  <li>Output enrichment score dataset</li>
</ol>

<p>A GCT file containing the input dataset’s projection onto a space of
specified gene sets. This GCT file may serve as input into
GenePattern’s many clustering and classification algorithms.</p>

<p>In the case of experimentally derived gene sets with _UP and _DN
suffixes appended to otherwise identical gene set names, <em>combine modes</em>
of <em>combine.add</em> and <em>combine.replace</em> will either add to the set or
replace the original gene set pair with a combined gene set with the
suffix removed from the name thereby creating new gene set names that
may impact downstream applications using these files in combination with
the original gene set collection file. Check that downstream
applications utilize subsets of gene sets within a collection for
compatibility with the <em>combine.add</em> mode output.</p>

<h2 id="platform-dependencies">Platform Dependencies</h2>

<p><strong>Task Type:</strong><br />
Projection</p>

<p><strong>CPU Type:</strong><br />
any</p>

<p><strong>Operating System:</strong><br />
any</p>

<p><strong>Language:</strong><br />
R-3.2</p>

<h2 id="version-comments">Version Comments</h2>

<ul>
  <li><strong>10.1.0</strong> (2022-10-2): Updated to Human MSigDB v2022.1.Hs and Mouse MSigDB 2022.1.Mm.</li>
  <li><strong>10.0.12</strong> (2022-9-15): Updated to Human MSigDB v2022.1.Hs. Direct support for Mouse MSigDB 2022.1.Mm is not yet available.</li>
  <li><strong>10.0.11</strong> (2022-1-19): Updated to MSigDB v7.5.1.</li>
  <li><strong>10.0.10</strong> (2022-1-12): Updated to MSigDB v7.5.</li>
  <li><strong>10.0.9</strong> (2021-4-22): Fixed minor typo.</li>
  <li><strong>10.0.8</strong> (2021-4-2): Updated to MSigDB v7.4.</li>
  <li><strong>10.0.7</strong> (2021-3-23): Updated to MSigDB v7.3.</li>
  <li><strong>10.0.6</strong> (2020-12-16): Fix for single sample GCT files and single set GMX files.</li>
  <li><strong>10.0.5</strong> (2020-12-15): Fix for single sample GCT files.</li>
  <li><strong>10.0.4</strong> (2020-9-23): Updated to MSigDB v7.2.</li>
  <li><strong>10.0.3</strong> (2020-4-2): Updated to MSigDB v7.1.</li>
  <li><strong>10.0.2</strong> (2020-3-10): Renamed from ssGSEAProjection. Updated to R-3.2. Documentation updates.</li>
  <li><strong>10.0.1</strong> (2020-2-26): Renamed from ssGSEAProjection. Updated to R-3.2.</li>
  <li><strong>9.1.2</strong> (2019-10-16): Updated to give access to MSigDB v7.0. Unified the Gene Set DB selector parameters and better downloading of MSigDB files. Removed Java dependency. Added a no-normalize option as the default.</li>
  <li><strong>8</strong> (2017-05-18): Updated to give access to MSigDB v6.0. Updated to use GenePattern native file lists. Fixes a bug to improve certain error messages.</li>
  <li><strong>7</strong> (2016-02-04): Updated to give access to MSigDB v5.1. Updated to R-2.15 and made fixes for portability.</li>
  <li><strong>6</strong> (2015-06-16): Add built-in support for MSigDB v5.0, which includes new hallmark gene sets.</li>
  <li><strong>5</strong> (2014-08-11): Added combine mode parameter.</li>
  <li><strong>4</strong> (2013-06-17): Updated list of gene sets to include v4.0 MSigDB collections.</li>
  <li><strong>3</strong> (2013-02-15): Updated list of gene sets databases to include v3.1 MSigDB collection, updated FTP download code, made documentation more biologist-friendly.</li>
  <li><strong>2</strong> (2012-09-19): Added support for reading of gmx-formatted gene set description files.</li>
  <li><strong>1</strong> (2012-07-03)</li>
</ul>

<p>Copyright © 2003-2022 Broad Institute, Inc., Massachusetts Institute of
Technology, and Regents of the University of California. All rights
reserved.</p>



      </section>
      <footer>
        
        <p><small>Theme adapted from <a href="https://twitter.com/michigangraham">mattgraham</a></small></p>
      </footer>
    </div>

    
  </body>
</html>