<div class="gp-content-header fluid">
<div class="container-fluid">
<div class="row">
<div class="col-md-12">
<h4><img src="./assets/gplogo.png"/>Module Documentation</h4>
<h1><big>STREAM.DetectTransitionGenes</big></h1>
</div>
</div>
</div>
</div><div class="container-fluid">
<div class="row">
<div class="col-md-12">
<br/>
<dl class="dl-horizontal">
<dt>LSID</dt>
<dd>urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00404</dd>
<dt>Author(s)</dt>
<dd>Huidong Chen, Massachussets General Hospital, wrapped as a module by Ted Liefeld, Mesirov Lab, UCSD School of Medicine.</dd>
<dt>Contact(s)</dt>
<dd><p>Algorithm and scientific questions: &lt;Huidong.Chen  at mgh dot harvard  dot edu&gt;</p>
<p>Module wrapping issues:  Ted Liefeld  &lt; jliefeld at cloud dot ucsd dot edu&gt;</p>
</dd>
</dl>
<hr/>
<h2>Introduction</h2>
<p>STREAM (Single-cell Trajectories Reconstruction, Exploration And Mapping) is an interactive pipeline capable of disentangling and visualizing complex branching trajectories from both single-cell transcriptomic and epigenomic data.  Within GenePattern STREAM is implemented as a collection of modules that cover the entire STREAM processing pipeline to allow individual steps to be performed interactively for data exploration.</p>
<p>STREAM.DetectTransitionGenes is used to detect marker genes for each transition.	</p>
<h2>Algorithm</h2>
<p>
For each branch Bi and for each gene E we first scale the gene expression values to [0,1] for convenience. Then we check if the candidate gene has a reasonable dynamic range considering cells close to the start and end points. To this end, we consider the fold change in average gene expressions of the first 20% and the last 80% of the cells based on the inferred pseudotime. If the difference is greater than a specified threshold (the default log2 fold change value is 0.25), we then calculate Spearman’s rank correlation between inferred pseudotime and gene expression of all the cells along Bi. Genes with Spearman’s correlation coefficient above a specified threshold (0.4 by default) are identified and reported as transition genes.</p>
<h2>References</h2>
<p>H Chen, L Albergante, JY Hsu, CA Lareau, GL Bosco, J Guan, S Zhou, AN Gorban, DE Bauer, MJ Aryee, DM Langenau, A Zinovyev, JD Buenrostro, GC Yuan, L Pinello <a href="https://www.nature.com/articles/s41467-019-09670-4">Single-cell trajectories reconstruction, exploration and mapping of omics data with STREAM. </a> Nature Communications, volume 10, Article number: 1903 (2019)</p>
<p>Nestorowa, S. et al.<a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425"> A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>Pinello Lab   <a href="https://github.com/pinellolab/STREAM">STREAM Github Repository</a></p>
<h2>Input Files</h2>
<ol>
<li>data file *<br/>
 A STREAM pkl file containing an annotated AnnData matrix of gene expression data.</li>
</ol>
<h2>Output Files</h2>
<ol>
<li>&lt;output filename&gt;_stream_result.pkl<br>
	Output file in STREAM AnnData extended pickle (.pkl) file format suitable for passing to the next step of the STREAM analysis.</br></li>
<li>transition_genes/transition_genes_S#_S#.tsv<br/>
        The transition genes for the specific  branching.  Columns include gene name, stat, log fold change, pval and qval.
        </li>
<li>transition_genes_S#_S#.pdf<br/>
        Bar plot showing Spearman correlation coefficient for genes between 2 branches..
        </li>
</ol>
<h2>Example Data</h2>
<p>Example data for the STREAM workflow can be downloaded from dropbox: <a href="https://www.dropbox.com/sh/dfqht1ob89ku99d/AACI5ZW3aRuq9MhBfSNS_1O_a?dl=0">Stream Example Data</a><br/>
Ref: Nestorowa, S. et al. <a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425">A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>Example data for this specific step can be found at <a href="https://datasets.genepattern.org/data/test_data/STREAM/stream_epg_result.pkl">stream_epg_result.pkl</a></p>
<h2>Requirements</h2>
<p>GenePattern 3.9.11 or later (dockerized).</p>
<h2>Parameters</h2>
<table class="table table-striped">
<tbody>
<tr><td colspan="2"><h4>Inputs and Outputs</h4></td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>data file*</td>
<td>A STREAM pkl file containing an annotated AnnData matrix of gene expression data/td&gt;
		</td></tr>
<tr>
<td>output filename*</td>
<td>The output filename prefix.</td>
</tr>
<tr><td colspan="2"><h4>Transition Gene Detection</h4>
			Parameters used if variable genes are to be selected as the feature.
                </td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>root   </td>
<td> The starting node.  </td>
</tr>
<tr>
<td> preference  </td>
<td> The preference of nodes. The branch with speficied nodes are preferred and put on the top part of subway plot. The higher ranks the node have, the closer to the top the branch with that node is. e.g. S3,S4.  </td>
</tr>
<tr>
<td>percentile expr*   </td>
<td> Between 0 and 100. Between 0 and 100. Specify the percentile of gene expression greater than 0 to filter out some extreme gene expressions.</td>
</tr>
<tr>
<td>use precomputed*   </td>
<td> 	If True, the previously computed scaled gene expression will be used.  </td>
</tr>
<tr>
<td> cutoff zscore*  </td>
<td> The z-score cutoff used for mean values of all leaf branches.  </td>
</tr>
<tr>
<td> cutoff pvalue  </td>
<td>  The p value cutoff used for Kruskal-Wallis H-test and post-hoc pairwise Conover's test. </td>
</tr>
<tr><td colspan="2"><h4>Plotting</h4>
                        Parameters controlling the output figures.
                </td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>num genes</td>
<td>The number of genes to plot </td>
</tr>
<tr>
<td>figure height/td&gt;
                        <td>Figure height as used in matplotlib graphs.  Default=8.</td>
</td></tr>
<tr>
<td>figure width</td>
<td>Figure width as used in matplotlib plots.  Default=8</td>
</tr>
</tbody>
</table>
<p><span style="color: red;">*</span> - required</p>
</div>
</div>
</div>