<div class="gp-content-header fluid">
<div class="container-fluid">
<div class="row">
<div class="col-md-12">
<h4><img src="./assets/gplogo.png"/>Module Documentation</h4>
<h1><big>STREAM.DimensionReduction</big></h1>
</div>
</div>
</div>
</div><div class="container-fluid">
<div class="row">
<div class="col-md-12">
<br/>
<dl class="dl-horizontal">
<dt>LSID</dt>
<dd>urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00396</dd>
<dt>Author(s)</dt>
<dd>Huidong Chen, Massachussets General Hospital, wrapped as a module by Ted Liefeld, Mesirov Lab, UCSD School of Medicine.</dd>
<dt>Contact(s)</dt>
<dd><p>Algorithm and scientific questions: &lt;Huidong.Chen  at mgh dot harvard  dot edu&gt;</p>
<p>Module wrapping issues:  Ted Liefeld  &lt; jliefeld at cloud dot ucsd dot edu&gt;</p>
</dd>
</dl>
<hr/>
<h2>Introduction</h2>
<p>STREAM (Single-cell Trajectories Reconstruction, Exploration And Mapping) is an interactive pipeline capable of disentangling and visualizing complex branching trajectories from both single-cell transcriptomic and epigenomic data.  Within GenePattern STREAM is implemented as a collection of modules that cover the entire STREAM processing pipeline to allow individual steps to be performed interactively for data exploration.</p>
<p>STREAM.DimensionReduction is used to reduce the dimensionality of the dataset to be used in the downstream analysis.</p>
<p> </p>
<h2>Algorithm</h2>
<p>
Each cell can be thought as a vector in a multi-dimensional vector space in which each component is the expression level of a gene. Typically, even after feature selection, each cell still has hundreds of components, making it difficult to reliably assess similarity or distances between cells, a problem often referred as the curse of dimensionality. To mitigate this problem, starting from the genes selected in the previous step we project cells to a lower dimensional space using a non-linear dimensionality reduction method called Modified Locally Linear Embedding (MLLE).
</p><p>
Several alternative dimension reduction methods are also supported, spectral embedding, umap, pca. By default, this module uses MLLE.
</p><p>
For large datasets, spectral embedding works faster than MLLE while preserving a similar compact structure to MLLE. For large datasets, lowering the percent neighbor cells parameter (0.1 by default) will speed up this step.
</p><p>
By default we set the number of components to keep to 3.  For biological process with simple bifurcation or linear trajectory, keeping only two components would be recommended.
</p>
<h2>References</h2>
<p>H Chen, L Albergante, JY Hsu, CA Lareau, GL Bosco, J Guan, S Zhou, AN Gorban, DE Bauer, MJ Aryee, DM Langenau, A Zinovyev, JD Buenrostro, GC Yuan, L Pinello <a href="https://www.nature.com/articles/s41467-019-09670-4">Single-cell trajectories reconstruction, exploration and mapping of omics data with STREAM. </a> Nature Communications, volume 10, Article number: 1903 (2019)</p>
<p>Nestorowa, S. et al.<a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425"> A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>Pinello Lab   <a href="https://github.com/pinellolab/STREAM">STREAM Github Repository</a></p>
<p><b>ADD GPNB NOTEBOOK HERE WHEN READY</b></p>
<h2>Input Files</h2>
<ol>
<li>data file *<br/>
 A STREAM pkl file containing an annotated AnnData matrix of gene expression data.</li>
</ol>
<h2>Output Files</h2>
<ol>
<li>&lt;output filename&gt;_stream_result.pkl<br>
	Output file in STREAM AnnData extended pickle (.pkl) file format suitable for passing to the next step of the STREAM analysis.</br></li>
<li>&lt;output filename&gt;_stddev_dotplot.png<bt></bt>
        Plot of the standard deviation of the dimensions. 
        </li>
</ol>
<h2>Example Data</h2>
<p>Example data can be downloaded from dropbox: <a href="https://www.dropbox.com/sh/dfqht1ob89ku99d/AACI5ZW3aRuq9MhBfSNS_1O_a?dl=0">Stream Example Data</a><br/>
Ref: Nestorowa, S. et al. <a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425">A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>Example data for the STREAM workflow can be downloaded from dropbox: <a href="https://www.dropbox.com/sh/dfqht1ob89ku99d/AACI5ZW3aRuq9MhBfSNS_1O_a?dl=0">Stream Example Data</a><br/>
Ref: Nestorowa, S. et al. <a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425">A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>An input file suitable for this step is available at <a href="https://datasets.genepattern.org/data/test_data/STREAM/preprocessed_stream_result.pkl">preprocessed_stream_result.pkl</a> </p>
<h2>Requirements</h2>
<p>GenePattern 3.9.11 or later (dockerized).</p>
<h2>Parameters</h2>
<table class="table table-striped">
<tbody>
<tr><td colspan="2"><h4>Inputs and Outputs</h4></td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>data file*</td>
<td>A STREAM pkl file containing an annotated AnnData matrix of gene expression data/td&gt;
		</td></tr>
<tr>
<td>output filename*</td>
<td>The output filename prefix.</td>
</tr>
<tr><td colspan="2"><h4>Dimension Reduction</h4>
</td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>percent neighbor cells   </td>
<td> The percentage neighbor cells (only valid when 'mlle', 'se', or 'umap' is specified).  </td>
</tr>
<tr>
<td>num components to keep</td>
<td>The number of components to keep in the resulting dataset.</td>
</tr>
<tr>
<td>feature</td>
<td>Feature used for dimension reduction. Choose from ['var_genes','top_pcs','all']. 'var_genes': most variable genes. 'top_pcs': top principal components. 'all': all genes.</td>
</tr>
<tr>
<td>method</td>
<td>Method used for dimension reduction.Choose from {{'mlle','umap','pca'}}</td>
</tr>

<tr><td colspan="2"><h4>Plotting</h4>
                        Parameters controlling the output figures.
                </td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>num components to plot</td>
<td>	Number of components to be plotted.</td>
</tr>
<tr>
<td>component x</td>
<td>	Component used for x axis.</td>
</tr>
<tr>
<td> component y</td>
<td>	Component used for y axis.</td>
</tr>
<tr>
<td>figure height</td>
<td>Figure height as used in matplotlib graphs.  Default=8.</td>
</tr>
<tr>
<td>figure width</td>
<td>Figure width as used in matplotlib plots.  Default=8</td>
</tr>
</tbody>
</table>
<p><span style="color: red;">*</span> - required</p>
</div>
</div>
</div>