<div class="gp-content-header fluid">
<div class="container-fluid">
<div class="row">
<div class="col-md-12">
<h4><img src="./assets/gplogo.png"/>Module Documentation</h4>
<h1><big>STREAM.SeedEPGStructure</big></h1>
</div>
</div>
</div>
</div><div class="container-fluid">
<div class="row">
<div class="col-md-12">
<br/>
<dl class="dl-horizontal">
<dt>LSID</dt>
<dd>urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00398</dd>
<dt>Author(s)</dt>
<dd>Huidong Chen, Massachussets General Hospital, wrapped as a module by Ted Liefeld, Mesirov Lab, UCSD School of Medicine.</dd>
<dt>Contact(s)</dt>
<dd><p>Algorithm and scientific questions: &lt;Huidong.Chen  at mgh dot harvard  dot edu&gt;</p>
<p>Module wrapping issues:  Ted Liefeld  &lt; jliefeld at cloud dot ucsd dot edu&gt;</p>
</dd>
</dl>
<hr/>
<h2>Introduction</h2>
<p>STREAM (Single-cell Trajectories Reconstruction, Exploration And Mapping) is an interactive pipeline capable of disentangling and visualizing complex branching trajectories from both single-cell transcriptomic and epigenomic data.  Within GenePattern STREAM is implemented as a collection of modules that cover the entire STREAM processing pipeline to allow individual steps to be performed interactively for data exploration.</p>
<p>STREAM.SeedEPGStructure is use to seed the initial elastic principal graph prior to starting the trajectory learning process.	</p>
<h2>Algorithm</h2>
<p>
Elastic principal graphs are structured data approximators, consisting of vertices connected by edges. The vertices are embedded into the space of the data, minimizing the mean squared distance (MSD) to the data points, similarly to k-means. Unlike unstructured k-means, the edges connecting the vertices are used to define an elastic energy term. The elastic energy term and MSD are used to create penalties for edge stretching and bending of branches.
</p>
<p>The principal graph inference is based on a greedy optimization procedure that may lead to local minima, therefore in STREAM we use the STREAM.SeedEPGStructure module as an initialization procedure that improves the quality of the inferred solutions and speeds up convergence. First, cells are clustered in the low-dimensional space (by default, k-means is used. Alternatively another two clustering methods including affinity propagation(ap) and spectral clustering(sc) are also available). Based on the centroids obtained, a minimum spanning tree (MST) is constructed using the Kruskal’s algorithm. The obtained tree is then used as initial tree structure for the ElPiGraph procedure.
</p>
<h2>References</h2>
<p>H Chen, L Albergante, JY Hsu, CA Lareau, GL Bosco, J Guan, S Zhou, AN Gorban, DE Bauer, MJ Aryee, DM Langenau, A Zinovyev, JD Buenrostro, GC Yuan, L Pinello <a href="https://www.nature.com/articles/s41467-019-09670-4">Single-cell trajectories reconstruction, exploration and mapping of omics data with STREAM. </a> Nature Communications, volume 10, Article number: 1903 (2019)</p>
<p>Nestorowa, S. et al.<a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425"> A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>Pinello Lab   <a href="https://github.com/pinellolab/STREAM">STREAM Github Repository</a></p>
<h2>Input Files</h2>
<ol>
<li>data file *<br/>
 A STREAM pkl file containing an annotated AnnData matrix of gene expression data.</li>
</ol>
<h2>Output Files</h2>
<ol>
<li>&lt;output filename&gt;_stream_result.pkl<br>
	Output file in STREAM AnnData extended pickle (.pkl) file format suitable for passing to the next step of the STREAM analysis.</br></li>
<li>&lt;output filename&gt;_variable_genes.png<bt></bt>
        Plot of genes against the fitted curve (if select variable genes is selected). 
        </li>
</ol>
<h2>Example Data</h2>
<p>Example data for the STREAM workflow can be downloaded from dropbox: <a href="https://www.dropbox.com/sh/dfqht1ob89ku99d/AACI5ZW3aRuq9MhBfSNS_1O_a?dl=0">Stream Example Data</a><br/>
Ref: Nestorowa, S. et al. <a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425">A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>An input file suitable for this step is available at <a href="https://datasets.genepattern.org/data/test_data/STREAM/dimred_stream_result.pkl">dimred_stream_result.pkl</a> </p>
<h2>Requirements</h2>
<p>GenePattern 3.9.11 or later (dockerized).</p>
<h2>Parameters</h2>
<table class="table table-striped">
<tbody>
<tr><td colspan="2"><h4>Inputs and Outputs</h4></td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>data file*</td>
<td>A STREAM pkl file containing an annotated AnnData matrix of gene expression data/td&gt;
		</td></tr>
<tr>
<td>output filename*</td>
<td>The output filename prefix.</td>
</tr>
<tr><td colspan="2"><h4>Structure Learning</h4>
</td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>percent neighbor cells*   </td>
<td> 	Neighbor percentage. The percentage of points used as neighbors for spectral clustering.  </td>
</tr>
<tr>
<td>num clusters*   </td>
<td>Number of clusters (only valid once 'clustering' is specificed as 'Spectral Clustering' or 'K-Means').   </td>
</tr>
<tr>
<td>damping*   </td>
<td>Damping factor (between 0.5 and 1) for affinity propagation.   </td>
</tr>
<tr>
<td>preference percentile*   </td>
<td>Preference percentile (between 0 and 100). The percentile of the input similarities for affinity propagation.   </td>
</tr>
<tr>
<td> max clusters*  </td>
<td> 	Number of clusters (only valid once 'clustering' is specificed as 'Spectral Clustering' or 'K-Means').  </td>
</tr>
<tr>
<td> clustering*  </td>
<td> Clustering method used to infer the initial nodes. Choose from affinity propagation, K-Means clustering, Spectral Clustering  </td>
</tr>
<tr>
<td> </td>
<td> </td>
</tr>
<tr><td colspan="2"><h4>Plotting</h4>
                        Parameters controlling the output figures.
                </td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>num components*   </td>
<td> The number of components to be plotted.  </td>
</tr>
<tr>
<td> component x*  </td>
<td> Component used for the x axis  </td>
</tr>
<tr>
<td> component y*  </td>
<td> Component used for the y axis  </td>
</tr>
<tr>
<td>figure height</td>
<td>Figure height as used in matplotlib graphs.  Default=8.</td>
</tr>
<tr>
<td>figure width</td>
<td>Figure width as used in matplotlib plots.  Default=8</td>
</tr>
<tr>
<td>figure legend num columns*   </td>
<td>	The number of columns that the legend has.   </td>
</tr>
</tbody>
</table>
<p><span style="color: red;">*</span> - required</p>
</div>
</div>
</div>