<div class="gp-content-header fluid">
<div class="container-fluid">
<div class="row">
<div class="col-md-12">
<h4><img src="./assets/gplogo.png"/>Module Documentation</h4>
<h1><big>STREAM.VisualizeTrajectories</big></h1>
</div>
</div>
</div>
</div><div class="container-fluid">
<div class="row">
<div class="col-md-12">
<br/>
<dl class="dl-horizontal">
<dt>LSID</dt>
<dd>urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00401</dd>
<dt>Author(s)</dt>
<dd>Huidong Chen, Massachussets General Hospital, wrapped as a module by Ted Liefeld, Mesirov Lab, UCSD School of Medicine.</dd>
<dt>Contact(s)</dt>
<dd><p>Algorithm and scientific questions: &lt;Huidong.Chen  at mgh dot harvard  dot edu&gt;</p>
<p>Module wrapping issues:  Ted Liefeld  &lt; jliefeld at cloud dot ucsd dot edu&gt;</p>
</dd>
</dl>
<hr/>
<h2>Introduction</h2>
<p>STREAM (Single-cell Trajectories Reconstruction, Exploration And Mapping) is an interactive pipeline capable of disentangling and visualizing complex branching trajectories from both single-cell transcriptomic and epigenomic data.  Within GenePattern STREAM is implemented as a collection of modules that cover the entire STREAM processing pipeline to allow individual steps to be performed interactively for data exploration.</p>
<p>STREAM.VisualizeTrajectories is used to generate the subway and stream plots for a computed graph structure. </p>
<p><h5>Huidong to add detail explaining this part of the STREAM pipeline</h5></p>
<p> </p>
<h2>Algorithm</h2>
<p>
<h5>Huidong to cover details of this portion of the analysis</h5>
</p>
<h2>References</h2>
<p>H Chen, L Albergante, JY Hsu, CA Lareau, GL Bosco, J Guan, S Zhou, AN Gorban, DE Bauer, MJ Aryee, DM Langenau, A Zinovyev, JD Buenrostro, GC Yuan, L Pinello <a href="https://www.nature.com/articles/s41467-019-09670-4">Single-cell trajectories reconstruction, exploration and mapping of omics data with STREAM. </a> Nature Communications, volume 10, Article number: 1903 (2019)</p>
<p>Nestorowa, S. et al.<a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425"> A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>Pinello Lab   <a href="https://github.com/pinellolab/STREAM">STREAM Github Repository</a></p>
<p><b>ADD GPNB NOTEBOOK HERE WHEN READY</b></p>
<h2>Input Files</h2>
<ol>
<li>data file *<br/>
 A STREAM pkl file containing an annotated AnnData matrix of gene expression data.</li>
</ol>
<h2>Output Files</h2>
<ol>
<li>&lt;output filename&gt;_stream_result.pkl<br>
	Output file in STREAM AnnData extended pickle (.pkl) file format suitable for passing to the next step of the STREAM analysis.</br></li>
<li>S#&lt;output filename&gt;_stream_plot_&lt;gene_name&gt;.png<bt></bt>
        Plot of gene expression across the named stream for the named gene. 
        </li>
<li>S#&lt;output filename&gt;_subwaymap_plot_&lt;gene_name&gt;.png<bt></bt>
        Subway map plot across the named stream for the named gene.           
        </li>
<li>&lt;output filename&gt;_flat_tree.png<bt></bt>
        Tree plot showing all requested gens across the trajectories.           
        </li>
</ol>
<h2>Example Data</h2>
<p>Example data for the STREAM workflow can be downloaded from dropbox: <a href="https://www.dropbox.com/sh/dfqht1ob89ku99d/AACI5ZW3aRuq9MhBfSNS_1O_a?dl=0">Stream Example Data</a><br/>
Ref: Nestorowa, S. et al. <a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425">A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>Example data for this specific module is available at <a href="https://datasets.genepattern.org/data/test_data/STREAM/stream_epg_result.pkl">stream_epg_result.pkl</a>.</p>
<h2>Requirements</h2>
<p>GenePattern 3.9.11 or later (dockerized).</p>
<h2>Parameters</h2>
<table class="table table-striped">
<tbody>
<tr><td colspan="2"><h4>General</h4></td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>data file*</td>
<td>A STREAM pkl file containing an annotated AnnData matrix of gene expression data/td&gt;
		</td></tr>
<tr>
<td>output filename*</td>
<td>The output filename prefix.</td>
</tr>
<tr>
<td>root   </td>
<td> The starting node.  </td>
</tr>
<tr>
<td>percentile dist   </td>
<td>Percentile of cells' distances from branches (between 0 and 100) used for calculating the distances between branches of subway map.   </td>
</tr>
<tr>
<td> preference  </td>
<td> The preference of nodes. The branch with speficied nodes are preferred and put on the top part of subway plot. The higher ranks the node have, the closer to the top the branch with that node is. e.g. S3,S4.  </td>
</tr>
<tr>
<td>color by   </td>
<td> 	Specify how to color cells.'label': the cell labels. 'branch': the branch id identifed by STREAM.  </td>
</tr>
<tr>
<td>subway factor  </td>
<td>The factor used to adjust the distances between branches of subway map.   </td>
</tr>
<tr>
<td>factor num win   </td>
<td> The number of sliding windows.  </td>
</tr>
<tr>
<td> factor min win  </td>
<td> The factor used to calculate the sliding window size based on shortest branch.  </td>
</tr>
<tr>
<td> factor width  </td>
<td> The ratio between length and width of stream plot.  </td>
</tr>
<tr><td colspan="2">
<h4>Cell Plots</h4>
</td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td> create cell plots  </td>
<td> Create the cell plots, True/False  </td>
</tr>
<tr>
<td>  log view </td>
<td>  If True,the number of cells (the width) is logarithmized when outputing stream plot. </td>
</tr>
<tr>
<td>factor zoom in   </td>
<td>	If "log view" parameter  is True, the factor used to zoom in the thin branches.   </td>
</tr>
<tr><td colspan="2">
<h4>Gene Plots</h4>
</td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>   create gene plots</td>
<td>  Create the gene plots, True/False. </td>
</tr>
<tr>
<td>genes   </td>
<td> 	Comma delimited list of genes, e.g. Car2,Gata1,Epx.  </td>
</tr>
<tr><td colspan="2"><h4>Plotting</h4>
                        Parameters controlling the output figures.
                </td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>figure height</td>
<td>Figure height as used in matplotlib graphs.  Default=8.</td>
</tr>
<tr>
<td>figure width</td>
<td>Figure width as used in matplotlib plots.  Default=8</td>
</tr>
<tr>
<td>figure legend num columns   </td>
<td> 	The number of columns that the legend has. </td>
</tr>
</tbody>
</table>
<p><span style="color: red;">*</span> - required</p>
</div>
</div>
</div>