<div class="gp-content-header fluid">
<div class="container-fluid">
<div class="row">
<div class="col-md-12">
<h4><img src="./assets/gplogo.png"/>Module Documentation</h4>
<h1><big>STREAM.Preprocess</big></h1>
</div>
</div>
</div>
</div><div class="container-fluid">
<div class="row">
<div class="col-md-12">
<br/>
<dl class="dl-horizontal">
<dt>LSID</dt>
<dd>urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00396</dd>
<dt>Author(s)</dt>
<dd>Huidong Chen, Massachussets General Hospital, wrapped as a module by Ted Liefeld, Mesirov Lab, UCSD School of Medicine.</dd>
<dt>Contact(s)</dt>
<dd><p>Algorithm and scientific questions: &lt;Huidong.Chen  at mgh dot harvard  dot edu&gt;</p>
<p>Module wrapping issues:  Ted Liefeld  &lt; jliefeld at cloud dot ucsd dot edu&gt;</p>
</dd>
</dl>
<hr/>
<h2>Introduction</h2>
<p>STREAM (Single-cell Trajectories Reconstruction, Exploration And Mapping) is an interactive pipeline capable of disentangling and visualizing complex branching trajectories from both single-cell transcriptomic and epigenomic data.  Within GenePattern STREAM is implemented as a collection of modules that cover the entire STREAM processing pipeline to allow individual steps to be performed interactively for data exploration.</p>
<p>STREAM.Preprocess is used to normalize and filter single-cell transcriptomic data and format it for analsysis using the STREAM piplene.</p>
<p>To prepare for processing in the follow-on STREAM modules, typically we will will first normalize the raw gene expression values based on library size. Then the gene expression values will be logarithmized. The mitochondrial genes will be removed.</p>
<h2>Algorithm</h2>
<p>
With this module, we can filter out cells based on several cell-centric metrics, including the minimum number of genes expressed, the minimum percentage of genes expressed, and the minimum number of read counts for one cell.
</p><p>
We can also filter out genes based on gene-centric metrics, including the minimum number of cells expressing one gene, the minimum percentage of cells expressing one gene, and the minimum number of read counts for one gene.
</p>
<h2>References</h2>
<p>H Chen, L Albergante, JY Hsu, CA Lareau, GL Bosco, J Guan, S Zhou, AN Gorban, DE Bauer, MJ Aryee, DM Langenau, A Zinovyev, JD Buenrostro, GC Yuan, L Pinello <a href="https://www.nature.com/articles/s41467-019-09670-4">Single-cell trajectories reconstruction, exploration and mapping of omics data with STREAM. </a> Nature Communications, volume 10, Article number: 1903 (2019)</p>
<p>Nestorowa, S. et al.<a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425"> A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>Pinello Lab   <a href="https://github.com/pinellolab/STREAM">STREAM Github Repository</a></p>
<p><b>ADD GPNB NOTEBOOK HERE WHEN READY</b></p>
<h2>Input Files</h2>
<ol>
<li>data file *<br/>
 A csv, tsv or STREAM pkl file containing single cell transcriptomic data.</li>
<li>cell label file<br>
        A tsv file containing cell labels.</br></li>
<li> cell label color file<br/>
        A tsv file containing colors (as hex values, e.g. #FF0000) to use for cell labels when creating stream plots.</li>
</ol>
<h2>Output Files</h2>
<ol>
<li>&lt;output filename&gt;_stream_result.pkl<br/>
	Output file in STREAM AnnData extended pickle (.pkl) file format suitable for passing to the next step of the STREAM analysis.</li>
</ol>
<h2>Example Data</h2>
<p>Example data can be downloaded from dropbox: <a href="https://www.dropbox.com/sh/dfqht1ob89ku99d/AACI5ZW3aRuq9MhBfSNS_1O_a?dl=0">Stream Example Data</a><br/>
Ref: Nestorowa, S. et al. <a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425">A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<h2>Requirements</h2>
<p>GenePattern 3.9.11 or later (dockerized).</p>
<h2>Parameters</h2>
<table class="table table-striped">
<tbody>
<tr><td colspan="2"><h4>Inputs and Outputs</h4></td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>data file*</td>
<td>A STREAM pkl file containing an annotated AnnData matrix of gene expression data/td&gt;
		</td></tr>
<tr>
<td>cell label file</td>
<td>	A tsv file containing cell labels.</td>
</tr>
<tr>
<td>cell label color file</td>
<td>A tsv file containing cell label colors</td>
</tr>
<tr>
<td>output filename*</td>
<td>The output filename prefix.</td>
</tr>
<tr><td colspan="2"><h4>Cell Filtering</h4>
</td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td> min percent genes  </td>
<td> 	The minimum percentage of genes expressed to keep a cell.  </td>
</tr>
<tr>
<td> min count genes  </td>
<td> The minimum number of read counts for each gene.  </td>
</tr>
<tr><td colspan="2">
<h4>Gene Filtering</h4>
</td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td> min num cells  </td>
<td> 	The minimum number of cells expressing a gene.  </td>
</tr>
<tr>
<td> min percent cells  </td>
<td> The minimum percentage of cells expressing a gene to keep a gene.  </td>
</tr>
<tr>
<td> min count cells  </td>
<td> The minimum number of read counts for one cell.  </td>
</tr>
<tr><td colspan="2"><h4>Other Preprocessing</h4>
</td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>expression cutoff &lt; /td&gt;
                        <td>	The expression cutoff used to determine if a gene is expressed. If expression is greater than expr_cutoff,the gene is considered 'expressed'.</td>
</td></tr>
<tr>
<td>normalize</td>
<td>Normalize the data, True/False</td>
</tr>
<tr>
<td>log transform</td>
<td>Log transform the dataset, True/False</td>
</tr>
<tr>
<td>remove mitochondrial genes</td>
<td>Remove mitochondrial genes, True/False</td>
</tr>
</tbody>
</table>
<p><span style="color: red;">*</span> - required</p>
</div>
</div>
</div>