<div class="gp-content-header fluid">
<div class="container-fluid">
<div class="row">
<div class="col-md-12">
<h4><img src="./assets/gplogo.png"/>Module Documentation</h4>
<h1><big>STREAM.ElasticPrincipalGraph</big></h1>
</div>
</div>
</div>
</div><div class="container-fluid">
<div class="row">
<div class="col-md-12">
<br/>
<dl class="dl-horizontal">
<dt>LSID</dt>
<dd>urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00399</dd>
<dt>Author(s)</dt>
<dd>Huidong Chen, Massachussets General Hospital, wrapped as a module by Ted Liefeld, Mesirov Lab, UCSD School of Medicine.</dd>
<dt>Contact(s)</dt>
<dd><p>Algorithm and scientific questions: &lt;Huidong.Chen  at mgh dot harvard  dot edu&gt;</p>
<p>Module wrapping issues:  Ted Liefeld  &lt; jliefeld at cloud dot ucsd dot edu&gt;</p>
</dd>
</dl>
<hr/>
<h2>Introduction</h2>
<p>STREAM (Single-cell Trajectories Reconstruction, Exploration And Mapping) is an interactive pipeline capable of disentangling and visualizing complex branching trajectories from both single-cell transcriptomic and epigenomic data.  Within GenePattern STREAM is implemented as a collection of modules that cover the entire STREAM processing pipeline to allow individual steps to be performed interactively for data exploration.</p>
<p>STREAM.ElasticPrincipalGraph is uses elastic principal graph learning to calculate a pseudotime trajectory. </p>
<p>Elastic principal graphs are structured data approximators, consisting of vertices connected by edges. The vertices are embedded into the space of the data, minimizing the mean squared distance (MSD) to the data points, similarly to k-means. Unlike unstructured k-means, the edges connecting the vertices are used to define an elastic energy term. The elastic energy term and MSD are used to create penalties for edge stretching and bending of branches. </p>
<h2>Algorithm</h2>
<p>
The STREAM.ElasticPrincipalGraph module uses the R-language <a href="https://github.com/Albluca/ElPiGraph.R">ElPiGraph</a> implementation of Elastic Principal Graphs.  To find the optimal graph structure, ElPiGraph uses a topological grammar (or, graph grammar) approach.  ElPiGraph is a completely redesigned algorithm for the previously introduced elastic principal graph optimization based on the use of elastic matrix Laplacian, trimmed mean square error, explicit control of topological complexity and scalability to millions of points on an ordinary laptop. 
</p>
<h2>References</h2>
<p>H Chen, L Albergante, JY Hsu, CA Lareau, GL Bosco, J Guan, S Zhou, AN Gorban, DE Bauer, MJ Aryee, DM Langenau, A Zinovyev, JD Buenrostro, GC Yuan, L Pinello <a href="https://www.nature.com/articles/s41467-019-09670-4">Single-cell trajectories reconstruction, exploration and mapping of omics data with STREAM. </a> Nature Communications, volume 10, Article number: 1903 (2019)</p>
<p>Nestorowa, S. et al.<a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425"> A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>Pinello Lab   <a href="https://github.com/pinellolab/STREAM">STREAM GitHub Repository</a></p>
<p>ElPiGraph <a href="https://github.com/Albluca/ElPiGraph.R"> GitHub Repository</a></p>.
<p><b>ADD GPNB NOTEBOOK HERE WHEN READY</b></p>
<h2>Input Files</h2>
<ol>
<li>data file *<br/>
 A STREAM pkl file containing an annotated AnnData matrix of gene expression data.</li>
</ol>
<h2>Output Files</h2>
<ol>
<li>&lt;output filename&gt;_stream_result.pkl<br>
	Output file in STREAM AnnData extended pickle (.pkl) file format suitable for passing to the next step of the STREAM analysis.</br></li>
<li>&lt;output filename&gt;_branches.png<bt></bt>
        Plot of showing the trajectory branches. 
        </li>
<li>&lt;output filename&gt;_branches_with_cells.png<bt></bt>
        Plot of showing the trajectory branches with the cells positioned on them.                                      
        </li>
</ol>
<h2>Example Data</h2>
<p>Example data for the complete STREAM workflow can be downloaded from dropbox: <a href="https://www.dropbox.com/sh/dfqht1ob89ku99d/AACI5ZW3aRuq9MhBfSNS_1O_a?dl=0">Stream Example Data</a><br/>
Ref: Nestorowa, S. et al. <a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425">A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>Example data for this module (this step in the workflow) is available as <a href="https://datasets.genepattern.org/data/test_data/STREAM/seeded_stream_result.pkl">seeded_stream_result.pkl</a></p>
<h2>Requirements</h2>
<p>GenePattern 3.9.11 or later (dockerized).</p>
<h2>Parameters</h2>
<table class="table table-striped">
<tbody>
<tr><td colspan="2"><h4>Inputs and Outputs</h4></td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>data file*</td>
<td>A STREAM pkl file containing an annotated AnnData matrix of gene expression data/td&gt;
		</td></tr>
<tr>
<td>output filename*</td>
<td>The output filename prefix.</td>
</tr>
<tr><td colspan="2"><h4>Elastic Principal Graph</h4>
</td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>epg num nodes*   </td>
<td> Number of nodes for elastic principal graph.  </td>
</tr>
<tr>
<td>incremental number of nodes*   </td>
<td> 	Incremental number of nodes for elastic principal graph when epg_n_nodes is not big enough.  </td>
</tr>
<tr>
<td> epg trimming radius*  </td>
<td>	Maximal distance from a node to the points it controls in the embedding.   </td>
</tr>
<tr>
<td>epg alpha*   </td>
<td> Alpha parameter of the penalized elastic energy.  </td>
</tr>
<tr>
<td>epg beta*   </td>
<td>Beta parameter of the penalized elastic energy.   </td>
</tr>
<tr>
<td>epg lambda*   </td>
<td>	Lambda parameter used to compute the elastic energy.   </td>
</tr>
<tr>
<td>epg mu*   </td>
<td> Mu parameter used to compute the elastic energy.  </td>
</tr>
<tr>
<td>epg final energy*   </td>
<td>	Indicate the final elastic energy associated with the configuration.   </td>
</tr>
<tr>
<td> </td>
<td> </td>
</tr>
<tr><td colspan="2"><h4>Plotting</h4>
                        Parameters controlling the output figures.
                </td></tr>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
<tr>
<td>num componenets   </td>
<td>The number of components to be plotted   </td>
</tr>
<tr>
<td>component x   </td>
<td>Component used for x-axis in plots   </td>
</tr>
<tr>
<td>component y   </td>
<td>Component used for y axis in plots   </td>
</tr>
<tr>
<td>figure height</td>
<td>Figure height as used in matplotlib graphs.  Default=8.</td>
</tr>
<tr>
<td>figure width</td>
<td>Figure width as used in matplotlib plots.  Default=8</td>
</tr>
<tr>
<td>figure legend num columns*   </td>
<td>The number of columns that the legend has.   </td>
</tr>
</tbody>
</table>
<p><span style="color: red;">*</span> - required</p>
</div>
</div>
</div>